pkg/osquery/tables/gsettings/metadata.go

```diff
@@ -0,0 +1,274 @@
+			level.Info(t.logger).Log(
+				"msg", "error describing key",
+				"key", k,
+				"schema", schema,
+				"err", err,
+			)
+			continue
+		}
+		descriptions = append(descriptions, desc)
+	}
+
+	return descriptions, nil
+}
+
+func (t *GsettingsMetadata) listKeys(ctx context.Context, schema string) ([]string, error) {
+	var keys []string
+	output := new(bytes.Buffer)
+
+	err := t.cmdRunner(ctx, []string{"list-keys", schema}, output)
+	if err != nil {
+		return keys, errors.Wrap(err, "fetching keys")
+	}
+	scanner := bufio.NewScanner(output)
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "" {
+			continue
+		}
+
+		keys = append(keys, line)
+	}
+
+	if err := scanner.Err(); err != nil {
+		level.Info(t.logger).Log("msg", "scanner error", "err", err)
+	}
+
+	return keys, nil
+}
+
+func (t *GsettingsMetadata) describeKey(ctx context.Context, schema, key string) (keyDescription, error) {
+	desc := keyDescription{Key: key}
+
+	d, err := t.getDescription(ctx, schema, key)
+	if err != nil {
+		return desc, errors.Wrap(err, "getting key's description")
+	}
+	desc.Description = d
+
+	datatype, err := t.getType(ctx, key, schema)
+	if err != nil {
+		return desc, errors.Wrap(err, "discerning key's type")
+	}
+	desc.Type = datatype
+
+	return desc, nil
+}
+
+func (t *GsettingsMetadata) getDescription(ctx context.Context, schema, key string) (string, error) {
+	output := new(bytes.Buffer)
+
+	err := t.cmdRunner(ctx, []string{"describe", schema, key}, output)
+	if err != nil {
+		return "", errors.Wrap(err, "describing key")
+	}
+
+	return strings.TrimSpace(output.String()), nil
+}
+
+func (t *GsettingsMetadata) getType(ctx context.Context, key, schema string) (string, error) {
+	output := new(bytes.Buffer)
+
+	err := t.cmdRunner(ctx, []string{"range", schema, key}, output)
+	if err != nil {
+		return "", errors.Wrap(err, "running 'gsettings range'")
+	}
+
+	result := strings.TrimSpace(strings.ReplaceAll(output.String(), "\n", " "))
+	// enum types need special formatting to distinguish the type (enum) from
+	// the possible values
+	if strings.HasPrefix(result, "enum") {
+		s := strings.TrimPrefix(result, "enum ")
+		vals := strings.Split(s, " ")
+		return fmt.Sprintf("enum: [ %s ]", strings.Join(vals, ",")), nil
+	}
+
+	// 'range' datatypes also need special handling
+	if strings.HasPrefix(result, "range ") {
+		s := strings.TrimPrefix(result, "range ")
+		parts := strings.Split(s, " ")
+
+		typ := convertType(parts[0])
+		var scope string
+		if len(parts) > 2 {
+			scope = fmt.Sprintf(" (%v to %v)", parts[1], parts[2])
+		}
+
+		return fmt.Sprintf("%s%s", typ, scope), nil
+	}
+
+	return convertType(result), nil
+}
+
+func execGsettingsCommand(ctx context.Context, args []string, output *bytes.Buffer) error {
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	command := args[0]
+	cmd := exec.CommandContext(ctx, gsettingsPath, args...)
+	dir, err := ioutil.TempDir("", fmt.Sprintf("osq-gsettings-%s", command))
```

Comment:
I like the tempdir, and I even like the tempdir down in the exec command (where it's needed). But the effect is that we're creating a huge number of them when this is called. Probably cleaner to create one in a parent and pass it along.